bitcoin/examples/sighash.rs

```diff
@@ -0,0 +1,238 @@
+        //Here we assume that we have an M of N multisig scriptPubKey.
+        for (k, instr) in script_sig.instructions().enumerate() {
+            match instr.unwrap() {
+                Instruction::PushBytes(pb) => {
+                    if k == 0 {
+                        assert!(
+                            pb.is_empty(),
+                            "first in ScriptSig must be PUSHBYTES_0 got {:?}",
+                            pb
+                        )
+                    } else if script_pubkey_bytes_opt.is_none()
+                        && k == script_sig.instructions().count() - 1
+                    {
+                        //In P2SH the last stack element from scriptSig is ScriptPubkey
+                        script_pubkey_bytes = pb.as_bytes();
+                    } else {
+                        //All others from 1 (0-indexed) to M-2 must be signatures
+                        let sig_bytes = pb.as_bytes();
+                        let sig = bitcoin::ecdsa::Signature::from_slice(sig_bytes)
+                            .expect("failed to parse sig");
+                        let sig_len = sig_bytes.len() - 1; //last byte is EcdsaSighashType sighash flag
+                                                           //ECDSA signature in DER format lengths are between 70 and 72 bytes
+                        assert!(
+                            (70..=72).contains(&sig_len),
+                            "signature length {} out of bounds",
+                            sig_len
+                        );
+                        //take sighash_flag into account - can they be different?
+                        assert!(
+                            last_sighash_flag == bitcoin::sighash::EcdsaSighashType::None
+                                || last_sighash_flag == sig.hash_ty,
+                            "different sighash flags"
+                        );
+                        last_sighash_flag = sig.hash_ty;
+                    }
+                    println!("ScriptSig PushBytes len {}: {:?}", pb.as_bytes().to_vec().len(), pb)
+                }
+                Instruction::Op(op) => {
+                    unreachable!("we only expect PushBytes here, got Op({})", op)
+                }
+            }
+        }
+    }
+
+    let script_code = bitcoin::Script::from_bytes(script_pubkey_bytes);
+    let cache = sighash::SighashCache::new(&tx);
+    let sighash = cache
+        .legacy_signature_hash(inp_idx, script_code, last_sighash_flag.to_u32())
+        .expect("failed to compute sighash");
+    println!("Legacy sighash: {:x}\n", sighash);
+}
+
+/// Computes sighash for a segwit multisig transaction input that spends a p2wsh output with "witness_v0_scripthash" scriptPubKey.type
+///
+/// # Arguments
+///
+/// * `raw_tx` - spending tx hex
+/// * `inp_idx` - spending tx input index
+/// * `value` - ref tx output value in sats
+fn compute_sighash_p2wsh(mut raw_tx: &[u8], inp_idx: usize, value: u64) {
```

Comment:
I'm not sure this is a generic "compute sighash for a p2wsh" input, it seems specific to a 2-of-2 multisig p2wsh because we make assumptions about the witness. Also if we do make these assumptions we can also make assumptions about the witness script and get the pubkeys out and verify the 2 signatures.

Reply:
> I'm not sure this is a generic "compute sighash for a p2wsh" input, it seems specific to a 2-of-2 multisig p2wsh because we make assumptions about the witness. Also if we do make these assumptions we can also make assumptions about the witness script and get the pubkeys out and verify the 2 signatures.

@tcharding In our earlier PR proposal, we had a general algorithm for a P2WSH M of N multisig signatures verifications for any values of M and N (and thus for any witness array size). I saved the code [here]( https://github.com/panicfarm/multisig-test/blob/d1107819870780ffb319f6e99183ef2ace786e9c/examples/sighash.rs#L304).
According @apoelstra [suggestion](https://github.com/rust-bitcoin/rust-bitcoin/pull/1776#issuecomment-1498346780), we removed all this logic from this PR, because Andrew only wanted to demonstrate the sighash computations in this example. We will be developing this code anyway, since we need it for our project. 

For now, we could merge this PR just with the sighash computations examples as is. Later we could add a separate verification [example file](https://github.com/panicfarm/multisig-test/blob/d1107819870780ffb319f6e99183ef2ace786e9c/examples/sighash.rs#L269). If you are interested, you could help convert our logic into miniscript (although for our purposes, we are OK with it as is). We are working on adding the P2TR cases too.